### PR TITLE
Add support for custom platform templates

### PR DESCRIPTION
**Pull request by rosen-vladimirov**
Add new option when adding platform: `--platform-template`. When it is used, CLI will use the specified template instead of the default template from the runtime. The path to the specified template will be saved in project's package.json, so next time when platform is added, the same template will be used.
In case when `--platform-template` is not passed, CLI will check the package.json and if there's value for the template, it will be used. Otherwise the default template from runtime will be used.

Sample usage:
```
$ tns create app1
$ cd app1
$ tns run android --platform-template https://github.com/rosen-vladimirov/android-platform-template/tarball/master
```

### CODE REVIEW

**Comment by rosen-vladimirov** [comment]
Original PR - https://github.com/NativeScript/nativescript-cli/pull/1525

File: lib/services/platform-service.ts [hunk 1]
```diff
@@ -114,7 +117,13 @@ export class PlatformService implements IPlatformService {
 			}
 
 			let sourceFrameworkDir = isFrameworkPathDirectory && this.$options.symlink ? path.join(this.$options.frameworkPath, "framework") : frameworkDir;
-			platformData.platformProjectService.createProject(path.resolve(sourceFrameworkDir), installedVersion).wait();
+			this.$projectDataService.initialize(this.$projectData.projectDir);
+			let customTemplateOptions = this.getPathToPlatformTemplate(this.$options.platformTemplate, platformData.frameworkPackageName).wait();
+			let pathToTemplate: string;
+			if(customTemplateOptions) {
```
Comment (rosen-vladimirov) [inline lib/services/platform-service.ts]: rewrite this: 
```TypeScript
let pathToTemplate = customTemplateOptions && customTemplateOptions.pathToTemplate;
```

File: lib/services/platform-service.ts [hunk 2]
```diff
@@ -132,12 +141,39 @@ export class PlatformService implements IPlatformService {
...
+				// read data from package.json's nativescript key
+				// check the nativescript.tns-<platform>.template value
+				let nativescriptPlatformData = this.$projectDataService.getValue(frameworkPackageName).wait();
+				selectedTemplate = nativescriptPlatformData && nativescriptPlatformData.template;
+			}
+
+			if(selectedTemplate) {
+				let tempDir = temp.mkdirSync("platform-template");
+				try {
+					let pathToTemplate = this.$npm.install(selectedTemplate, tempDir, {"ignore-scripts": true, "production": true}).wait()[0][1];
```
Comment (Mitko-Kerezov) [inline lib/services/platform-service.ts]: Please add a comment explaining why you get `[0][1]` from the `$npm.install`

**Comment by Mitko-Kerezov** [comment]
:+1: after addressing comments

**Comment by ikoevska** [comment]
Documentation? :D

**Comment by rosen-vladimirov** [comment]
@ikoevska thanks, I've updated the PR, can you check the documentation

File: docs/man_pages/project/configuration/platform-add.md [hunk 3]
```diff
@@ -3,26 +3,27 @@ platform add
...
 
 ### Options
-* `--frameworkPath` - Sets the path to a NativeScript runtime for the specified platform that you want to use instead of the default runtime. If `--symlink` is specified, `<File Path>` must point to directory in which the runtime is already extracted. If `--symlink` is not specified, `<File Path>` must point to a valid npm package. 
+* `--frameworkPath` - Sets the path to a NativeScript runtime for the specified platform that you want to use instead of the default runtime. If `--symlink` is specified, `<File Path>` must point to directory in which the runtime is already extracted. If `--symlink` is not specified, `<File Path>` must point to a valid npm package.
 * `--symlink` - Creates a symlink to a NativeScript runtime for the specified platform that you want to use instead of the default runtime. If `--frameworkPath` is specified, creates a symlink to the specified directory. If `--frameworkPath` is not specified, creates a symlink to platform runtime installed with your current version of NativeScript.
 * `--sdk` - Sets the target Android SDK for the project.
```
Comment (ikoevska) [inline docs/man_pages/project/configuration/platform-add.md]: You need to add a line for --platformTemplate

File: docs/man_pages/project/configuration/platform-add.md [hunk 4]
```diff
@@ -3,26 +3,27 @@ platform add
...
 
 ### Attributes
 * `<API Level>` is a valid Android API level. For example: 17, 19, MNC.<% if(isHtml) { %> For a complete list of the Android API levels and their corresponding Android versions, click [here](http://developer.android.com/guide/topics/manifest/uses-sdk-element.html#platform).<% } %>
 * `<File Path>` is the complete path to a valid npm package or a directory that contains a NativeScript runtime for the selected platform.
-* `<Version>` is any available version of the respective platform runtime published in npm. <% if(isHtml) { %>If `@<Version>` is not specified, the NativeScript CLI installs the latest stable runtime for the selected platform.  
-To list all available versions for Android, run `$ npm view tns-android versions`  
-To list only experimental versions for Android, run `$ npm view tns-android dist-tags`  
-To list all available versions for iOS, run `$ npm view tns-ios versions`  
-To list only experimental versions for iOS, run `$ npm view tns-ios dist-tags`  
+* `<Platform Template Path>` is the path to a valid npm package, directory, .tgz or github URL that contains native Android/iOS template.
```
Comment (ikoevska) [inline docs/man_pages/project/configuration/platform-add.md]: GitHub

File: docs/man_pages/project/configuration/platform-add.md (hunk 3 above)
Comment (rosen-vladimirov) [inline docs/man_pages/project/configuration/platform-add.md]: done

**Comment by rosen-vladimirov** [comment]
retarget to release for 1.7.0 ...

**Pull request closed without merging**